### PR TITLE
Fix:  Broken URL of Customizer with autofocus query string parameter

### PR DESCRIPTION
**Pull request by Debarghya-Banerjee**
**Trac Ticket - [Core-61561](https://core.trac.wordpress.org/ticket/61561)**

## **Summary**

Fixes an issue with the autofocus query string parameter in the WordPress Customizer (customize.php). Previously, sanitize_text_field() was used on $_REQUEST['autofocus'], which is an array, resulting in an empty string.

### **Problem**

The Customizer's autofocus functionality was broken due to incorrect sanitization of the $_REQUEST['autofocus'] variable. When sanitize_text_field() is used on an array, it returns an empty string.

### **Example**

```
customize.php?autofocus%5Bsection%5D=colors
```

### **Solution**

To resolve this issue, the keys and values of the $_REQUEST['autofocus'] array are sanitized separately and then combined into a new array.

### **Changes Made**

- Sanitize the keys of the $_REQUEST['autofocus'] array using sanitize_key().
- Sanitize the values of the $_REQUEST['autofocus'] array using sanitize_text_field().
- Combine the sanitized keys and values into a new array using array_combine().

### CODE REVIEW

**Comment by swissspidy** [comment]
@hellofromtonya why me? 🤔

**Comment by swissspidy** [comment]
Ah, because of https://core.trac.wordpress.org/changeset/58069. Context would have been helpful.

File: src/wp-admin/customize.php
Comment (swissspidy) [inline src/wp-admin/customize.php]: We don't really need to sanitize the keys as `WP_Customize_Manager::set_autofocus()` has an allowlist of keys: `panel`, `section`, `control`

So this should suffice:

```php
$autofocus = ! empty( $_REQUEST['autofocus'] ) && is_array( $_REQUEST['autofocus'] ) ? array_map( 'sanitize_text_field', wp_unslash( $_REQUEST['autofocus'] ) ) : array();
```

Also just noticing the other `wp_unslash` calls further below are incorrect, they should be happening before the sanitizing.

We can thus change the whole code section to this:

```php
$url       = ! empty( $_REQUEST['url'] ) ? sanitize_text_field( wp_unslash( $_REQUEST['url'] ) ) : '';
$return    = ! empty( $_REQUEST['return'] ) ? sanitize_text_field( wp_unslash( $_REQUEST['return'] ) ) : '';
$autofocus = ! empty( $_REQUEST['autofocus'] ) && is_array( $_REQUEST['autofocus'] )
		? array_map( 'sanitize_text_field', wp_unslash( $_REQUEST['autofocus'] ) )
		: array();

if ( ! empty( $url ) ) {
	$wp_customize->set_preview_url( $url );
}
if ( ! empty( $return ) ) {
	$wp_customize->set_return_url( $return );
}
if ( ! empty( $autofocus ) ) {
	$wp_customize->set_autofocus( $autofocus );
}
```

Comment (Debarghya-Banerjee) [reply src/wp-admin/customize.php]: Fixed

**Comment by Debarghya-Banerjee** [comment]
Hi @swissspidy , I have addressed the feedback and made the necessary changes. Assigning you again for re-review.

**Review by swissspidy — APPROVED** [review]

File: src/wp-admin/customize.php
Comment (hellofromtonya) [inline src/wp-admin/customize.php]: nit for readability:

```suggestion
$autofocus = ! empty( $_REQUEST['autofocus'] ) && is_array( $_REQUEST['autofocus'] )
	? array_map( 'sanitize_text_field', wp_unslash( $_REQUEST['autofocus'] ) )
	: array(); // Since `$_REQUEST['autofocus']` is an array.
```

File: src/wp-admin/customize.php
Comment (ironprogrammer) [inline src/wp-admin/customize.php]:
```suggestion
$url       = ! empty( $_REQUEST['url'] ) ? sanitize_text_field( wp_unslash( $_REQUEST['url'] ) ) : '';
$return    = ! empty( $_REQUEST['return'] ) ? sanitize_text_field( wp_unslash( $_REQUEST['return'] ) ) : '';
$autofocus = ! empty( $_REQUEST['autofocus'] ) && is_array( $_REQUEST['autofocus'] )
```
Suggested for [coding standards](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#indentation).

File: src/wp-admin/customize.php
Comment (ironprogrammer) [inline src/wp-admin/customize.php]:
```suggestion
	: array(); // `$_REQUEST['autofocus']` is an array.
```
"Since" could be confused with `@since`, so this word might be omitted, or changed to "Because" to remove [ambiguity](https://grammarbrain.com/because-vs-since/).

**Review by ironprogrammer — COMMENTED** [review]
Thanks for the PR, @Debarghya-Banerjee! Here are a couple small suggestions for you to consider.

File: src/wp-admin/customize.php
Comment (hellofromtonya) [reply src/wp-admin/customize.php]: Good point @ironprogrammer.

Thinking this inline comment isn't necessary, as it's self-explanatory. 

How so?

It's setting a default value of an empty `array`. When evaluating the `true` branch of the ternary, an `array` will be assigned to `$autofocus` (via `array_map()` - see [PHP docs](https://www.php.net/manual/en/function.array-map.php)). Given that context and the default value, I'm thinking it's self-explanatory.

Comment (hellofromtonya) [inline src/wp-admin/customize.php]:
```suggestion
	: array();
```

File: src/wp-admin/customize.php
Comment (hellofromtonya) [inline src/wp-admin/customize.php:91]: @Debarghya-Banerjee why are the `wp_unslash()` being relocated to here? 

They were introduced long ago in [r34269](https://core.trac.wordpress.org/changeset/34269), not in [r58069](https://core.trac.wordpress.org/changeset/58069). 

I'm asking, as the focus of the PR is on restoring the `array` (which [r58069](https://core.trac.wordpress.org/changeset/58069) changed it to `string` in 6.6.0).

Comment (hellofromtonya) [reply src/wp-admin/customize.php:91]: IMO the `wp_unslash()` should not be changed in this fix for it to ship in 6.6.2 (given it's not part of the bug that was introduced in 6.6).

**Review by hellofromtonya — DISMISSED** [review]
I think this fix should narrowly focus on restoring autofocus from `string` (which was changed in [r58069](https://core.trac.wordpress.org/changeset/58069)) to `array`. I don't think relocating the `wp_unsplash()` is part of the regression fix.

**Comment by swissspidy** [comment]
@hellofromtonya As I suggested that change, I disagree. https://core.trac.wordpress.org/changeset/58069 made an incorrect change in that it a) called `sanitize_text_field` on an array and b) called `sanitize_text_field` before calling `wp_unslash()` (rather than after). Both are introduced in the 6.6 cycle and it makes sense to fix this properly, _together_

**Comment by Debarghya-Banerjee** [comment]
Hi @hellofromtonya , what should be done, should I revert back the `wp_unsplash()` changes?

File: src/wp-admin/customize.php
Comment (Debarghya-Banerjee) [reply src/wp-admin/customize.php:91]: Hi @hellofromtonya , what should be done, should I revert back the `wp_unsplash()` changes?

**Comment by hellofromtonya** [comment]
>As I suggested that change, I disagree. https://core.trac.wordpress.org/changeset/58069 made an incorrect change in that it a) called sanitize_text_field on an array and b) called sanitize_text_field before calling `wp_unslash()` (rather than after). Both are introduced in the 6.6 cycle and it makes sense to fix this properly, together

I didn't previously see that @swissspidy [suggested the change to relocate `wp_unslash()`](https://github.com/WordPress/wordpress-develop/pull/6974#discussion_r1682316457) to occur _before_ sanitizing:
>Also just noticing the other `wp_unslash` calls further below are incorrect, they should be happening before the sanitizing.

Noting, Core uses both approaches of `wp_unslash()` before and after sanitizing. For example, in `wp-signup.php`, `$_POST['WPLANG']` is sanitized first and then unslashed.

What about this specific bugfix, which is better?
Unsplashing _before_ does not contribute to the bug itself. That said, I agree with @swissspidy that it's better to unsplash before sanitizing. 

The reason for me asking to revert that change is to minimize the code changes to just the bug itself for the 6.6.2 minor cycle. But with this new understanding @swissspidy shared, I withdraw that suggestion and will move forward with committing the change.

Thank you everyone for your patience and contributions.

File: src/wp-admin/customize.php
Comment (hellofromtonya) [reply src/wp-admin/customize.php:91]: Sorry for my delay @Debarghya-Banerjee in responding. No changes are needed as I've withdrawn my suggestion (see https://github.com/WordPress/wordpress-develop/pull/6974#issuecomment-2248340365).

**Comment by hellofromtonya** [comment]
Committed via https://core.trac.wordpress.org/changeset/58804. Thank you everyone for your contributions ⭐

**Comment by Debarghya-Banerjee** [comment]
Hi @hellofromtonya , it seems that my ID is missing from the props. Can you please take a look into it. Thanks :-)

**Comment by hellofromtonya** [comment]
>it seems that my ID is missing from the props. Can you please take a look into it. Thanks :-)

Yeah I saw that after I committed. Sorry about that.

As I noted [here in the Trac ticket](https://core.trac.wordpress.org/ticket/61561#comment:22), I manually added you in Core's Props and noted for the committer who does the backport to 6.6 branch to add you to the commit `props`.

**Comment by ironprogrammer** [comment]
Hi, @Debarghya-Banerjee -- You can refer to the "Unlinked Props" section of [this comment](https://github.com/WordPress/wordpress-develop/pull/6974#issuecomment-2210516205), which has instructions to make sure you're automatically properly credited in the future. And thanks again for the PR!

**Comment by Debarghya-Banerjee** [comment]
> Hi, @Debarghya-Banerjee -- You can refer to the "Unlinked Props" section of [this comment](https://github.com/WordPress/wordpress-develop/pull/6974#issuecomment-2210516205), which has instructions to make sure you're automatically properly credited in the future. And thanks again for the PR!

Hi @ironprogrammer , I actually had some issues, but before the PR got approved, I fixed it, you can check my ID on the props as well. Thanks for youe response!

**Pull request closed without merging**
